src/Haskell/Ide/Engine/Dispatcher.hs

```diff
@@ -72,14 +72,20 @@ ideDispatcher env errorHandler callbackHandler pin = forever $ do
   (IdeRequest lid callback action) <- liftIO $ atomically $ readTChan pin
   debugm $ "ideDispatcher:got request with id: " ++ show lid
   cancelled <- liftIO $ atomically $ isCancelled env lid
-  unless cancelled $ do
-    response <- action
-    handleResponse lid callback response
+  if cancelled
+    then liftIO $ errorHandler lid J.RequestCancelled ""
+    else do
+      response <- action
+      handleResponse lid callback response
       
-  where handleResponse lid callback response = case response of
-          IdeResponseResult (IdeResultOk x) -> liftIO $ callbackHandler callback x
-          IdeResponseResult (IdeResultFail err) -> liftIO $ errorHandler lid err 
-          IdeResponseDeferred fp cacheCb -> handleDeferred lid fp cacheCb callback
+  where handleResponse lid callback response = do
+          cancelled <- liftIO $ atomically $ isCancelled env lid
+          if cancelled
+            then liftIO $ errorHandler lid J.RequestCancelled ""
+          else case response of
```

Comment:
This if statement is just re-doing the logic above, wrt cancelling. This function should just be the else leg